### PR TITLE
feature: header plugin

### LINKED ISSUE #65: dm-core-packages(dm-core): move header into separate plugin

### PR DESCRIPTION
**Pull request by soofstad**
## What does this pull request change?
- Adds the "header" plugin

## Why is this pull request needed?
- Provide an application header as a plugin

## Issues related to this change
closes #65

NOTE: The attribute/grid selector is disabled in this first version of the plugin

### CODE REVIEW

File: packages/header/src/Header.tsx [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+        {/*)}*/}
+      </TopBar.Header>
+      <TopBar.Actions>
+        <Icons>
+          <ClickableIcon onClick={() => setAboutOpen(true)}>
+            <Icon name="info_circle" size={24} title="About" />
+          </ClickableIcon>
+          <ClickableIcon onClick={() => setVisibleUserInfo(true)}>
+            <Icon name="account_circle" size={24} title="User" />
+          </ClickableIcon>
```
Comment (ingeridhellen) [inline packages/header/src/Header.tsx:144]: Shouldn't these only be shown if userInfo and about is set in THeaderPluginConfig?

File: packages/header/src/Header.tsx [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        </Icons>
+      </TopBar.Actions>
+      <Dialog
```
Comment (ingeridhellen) [inline packages/header/src/Header.tsx:147]: There is a TopBar.CustomContent that should be wrapped around this, I think

Comment (ingeridhellen) [inline packages/header/src/Header.tsx:147]: This syntax seems outdated. Perhaps update eds?

File: packages/header/src/Header.tsx [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        {/*    ))}*/}
+        {/*    <Link to={'/DMT/search'}>*/}
+        {/*      <AppBox>Search</AppBox>*/}
+        {/*    </Link>*/}
+        {/*  </AppSelectorWrapper>*/}
+        {/*)}*/}
+      </TopBar.Header>
+      <TopBar.Actions>
+        <Icons>
+          <ClickableIcon onClick={() => setAboutOpen(true)}>
```
Comment (ingeridhellen) [inline packages/header/src/Header.tsx:139]: I miss logic solving the issue of both dialogs being open at once.

File: packages/header/package.json [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    "dist",
+    "src",
+    "tsconfig.json"
+  ],
+  "main": "dist/index.js",
+  "types": "dist/index.d.ts",
+  "scripts": {
+    "prebuild": "rm -rf dist && yarn install",
+    "build": "tsc",
+    "postbuild": "cp -R blueprints package.json README.md ./dist/ && rm -rf node_modules"
```
Comment (ingeridhellen) [inline packages/header/package.json:30]:
```suggestion
    "postbuild": "cp -R package.json README.md ./dist/ && rm -rf node_modules"
```

File: packages/header/src/useLocalStorage.tsx [hunk 5]
```diff
@@ -0,0 +1,29 @@
+import { useState } from 'react'
+
+export function useLocalStorage<T>(
```
Comment (ingeridhellen) [inline packages/header/src/useLocalStorage.tsx:3]: In mercury, this file has been adapted to include versioning of the local storage, making it much easier to give nice error messages if the data in someones local storage is no longer compatible with our code

**Review by ingeridhellen — CHANGES_REQUESTED** [review]
I didn't succeed in testing this locally (running npm install && npm run build in all packages gave me several errors) so a bit hard to see how it will look

File: packages/header/src/Header.tsx (hunk 3 above)
Comment (soofstad) [reply packages/header/src/Header.tsx:139]: Clicking outside the Dialog closes the first one automatically

File: packages/header/src/useLocalStorage.tsx (hunk 5 above)
Comment (soofstad) [reply packages/header/src/useLocalStorage.tsx:3]: IMO, this should be a generic localStorage hook. And any incompatible data stored there should be handled by which ever component is calling it.

File: packages/header/src/Header.tsx (hunk 2 above)
Comment (soofstad) [reply packages/header/src/Header.tsx:147]: Updated EDS. Still seems correct. Also according to EDS docs.

Comment (soofstad) [reply packages/header/src/Header.tsx:147]: Wrapped with that. Don't think it changed anything :+1:

File: packages/header/src/Header.tsx (hunk 1 above)
Comment (soofstad) [reply packages/header/src/Header.tsx:144]: fixed

File: packages/header/src/Header.tsx [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+
+  if (isLoading || !entity) {
+    return <Loading />
+  }
+
+  return (
+    <TopBar>
+      <TopBar.Header>
+        {/*<ClickableIcon onClick={() => setAppSelectorOpen(!appSelectorOpen)}>*/}
+        <Icon name="grid_on" size={32} />
```
[the remark this answers is not in the document]
Comment (soofstad) [reply packages/header/src/Header.tsx:121]: Agreed. Ported as it was. But should use this instead.

File: packages/header/src/Header.tsx [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+const UnstyledList = styled.ul`
+  margin: 0;
+  padding: 0;
+  list-style-type: none;
+  display: inline-flex;
+`
+
+type THeaderPluginConfig = {
+  userInfo: boolean
+  about: boolean
```
[the remark this answers is not in the document]
Comment (soofstad) [reply packages/header/src/Header.tsx:86]: fixed

File: packages/header/package.json [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  "dependencies": {
+    "@equinor/eds-core-react": "^0.20.0",
+    "@equinor/eds-icons": "^0.12.0",
+    "@equinor/eds-tokens": "^0.7.0",
+    "@development-framework/dm-core": ">=1.0.35"
+  },
+  "devDependencies": {
+    "@types/react": "^17.0.39",
+    "@types/styled-components": "^4.1.18",
+    "typescript": "^4.5.5"
```
[the remark this answers is not in the document]
Comment (soofstad) [reply packages/header/package.json:14]: updated

File: packages/header/jest.config.js [hunk 9]
```diff
@@ -0,0 +1,8 @@
+const base = require('../../../jest.config.base')
```
[the remark this answers is not in the document]
Comment (soofstad) [reply packages/header/jest.config.js:1]: fixed

**Review by eoaksnes — APPROVED** [review]

File: packages/header/src/useLocalStorage.tsx (hunk 5 above)
Comment (ingeridhellen) [reply packages/header/src/useLocalStorage.tsx:3]: It would still be a generic hook. https://github.com/equinor/mercury/pull/289/files

Comment (ingeridhellen) [reply packages/header/src/useLocalStorage.tsx:3]: The problem is that js doesn't raise an error if the value returned by useLocalStorage does not fit with the type expected by the caller. Without versioning, the caller would have to check that the returned value, which could be quite complex, in fact has all the parameters expected by the type. Failure to add such a test when updating the type could cause failures anywhere in the code

**Pull request merged**